ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3MultipartCompleteUploadHandler.java

```diff
@@ -364,30 +363,28 @@ List<ChunkInfo> getChunksToStitch(CompleteMultipartUpload completeMultipartUploa
         Collections.sort(sortedParts, Comparator.comparingInt(Part::getPartNumber));
         String reservedMetadataId = null;
         for (Part part : sortedParts) {
-          PutBlobMetaInfo putBlobMetaInfoObj = PutBlobMetaInfo.deserialize(part.geteTag());
-          // reservedMetadataId can be null. but if it's not null, they are supposed to be the same
-          String reserved = putBlobMetaInfoObj.getReservedMetadataChunkId();
-          if (reservedMetadataId == null) {
-            reservedMetadataId = reserved;
-          }
-          if (reserved != null && !reserved.equals(reservedMetadataId)) {
-            String error = "Reserved ID are different " + completeMultipartUpload;
-            throw new RestServiceException(error, RestServiceErrorCode.BadRequest);
-          }
+          S3MultipartETag eTag = S3MultipartETag.deserialize(part.geteTag());
           // TODO [S3]: decide the life cycle of S3.
           long expirationTimeInMs = -1;
 
-          List<Pair<String, Long>> chunks = putBlobMetaInfoObj.getOrderedChunkIdSizeList();
-          for (int i = 0; i < putBlobMetaInfoObj.getNumChunks(); i++) {
-            String blobId = chunks.get(i).getFirst();
-            long chunkSize = chunks.get(i).getSecond();
+          if (eTag.getVersion() == S3MultipartETag.VERSION_1) {
+            List<Pair<String, Long>> chunks = eTag.getOrderedChunkIdSizeList();
+            for (int i = 0; i < chunks.size(); i++) {
+              String blobId = chunks.get(i).getFirst();
+              long chunkSize = chunks.get(i).getSecond();
 
-            ChunkInfo chunk = new ChunkInfo(blobId, chunkSize, expirationTimeInMs, reservedMetadataId);
-            chunkInfos.add(chunk);
+              ChunkInfo chunk = new ChunkInfo(blobId, chunkSize, expirationTimeInMs, reservedMetadataId);
+              chunkInfos.add(chunk);
+            }
+          } else {
+            String error = "Wrong ETag version " + completeMultipartUpload + " rest request " + restRequest;
+            LOGGER.error(error);
+            throw new RestServiceException(error, RestServiceErrorCode.BadRequest);
```

Comment:
Would we need `if and else` condition here? When ETag version is wrong, `S3MultipartETag.deserialize()` throws exception and we might directly go to `catch` block.

Reply:
That's true. 
But with the else, I think it increases the readability. And if we have version 2 and S3MultipartETag can deserialize it,  we won't miss the case in the handler code.